internal/pkg/otel/manager/manager.go

```diff
@@ -264,20 +299,221 @@ func (m *OTelManager) Errors() <-chan error {
 	return m.errCh
 }
 
-// Update updates the configuration.
-//
-// When nil is passed for the cfg, then the collector is stopped.
-func (m *OTelManager) Update(cfg *confmap.Conf) {
+// handleCollectorUpdate processes collector configuration updates received through the collectorUpdateCh.
+// This method updates the internal collector configuration and triggers a rebuild of the merged
+// configuration that combines collector and component configurations.
+func (m *OTelManager) handleCollectorUpdate(cfg *confmap.Conf) error {
+	m.collectorCfg = cfg
+	return m.updateMergedConfig()
+}
+
+// handleComponentUpdate processes component model updates received through the componentUpdateCh.
+// This method updates the internal component list and triggers a rebuild of the merged
+// configuration that combines collector and component configurations.
+func (m *OTelManager) handleComponentUpdate(model component.Model) error {
+	m.components = model.Components
+	return m.updateMergedConfig()
+}
+
+// buildMergedConfig combines collector configuration with component-derived configuration.
+func (m *OTelManager) buildMergedConfig() (*confmap.Conf, error) {
+	mergedOtelCfg := confmap.New()
+
+	// Generate component otel config if there are components
+	var componentOtelCfg *confmap.Conf
+	if len(m.components) > 0 {
+		model := &component.Model{Components: m.components}
+		var err error
+		componentOtelCfg, err = translate.GetOtelConfig(model, m.agentInfo, m.beatMonitoringConfigGetter)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate otel config: %w", err)
+		}
+	}
+
+	// If both configs are nil, return nil so the manager knows to stop the collector
+	if componentOtelCfg == nil && m.collectorCfg == nil {
+		return nil, nil
+	}
+
+	// Merge component config if it exists
+	if componentOtelCfg != nil {
+		err := mergedOtelCfg.Merge(componentOtelCfg)
+		if err != nil {
+			return nil, fmt.Errorf("failed to merge component otel config: %w", err)
+		}
+	}
+
+	// Merge with base collector config if it exists
+	if m.collectorCfg != nil {
+		err := mergedOtelCfg.Merge(m.collectorCfg)
+		if err != nil {
+			return nil, fmt.Errorf("failed to merge collector otel config: %w", err)
+		}
+	}
+
+	return mergedOtelCfg, nil
+}
+
+// updateMergedConfig builds the merged configuration for the otel manager by merging the base collector configuration
+// with the component configuration, and updates the otel manager with the merged configuration.
+func (m *OTelManager) updateMergedConfig() error {
+	mergedCfg, err := m.buildMergedConfig()
+	if err != nil {
+		return err
+	}
+
+	m.collectorConfigMutex.Lock()
+	defer m.collectorConfigMutex.Unlock()
+	m.mergedCollectorCfg = mergedCfg
+	return nil
+}
+
+func (m *OTelManager) applyMergedConfig(ctx context.Context, collectorStatusCh chan *status.AggregateStatus, collectorRunErr chan error) error {
+	if m.proc != nil {
```

Comment:
I am not a fun of proc getting migrated to the OtelManager struct, since this is purely scoped to the Run function. So how about making this a private function where you pass the manager and proc as parameters? 🙂

Reply:
I tried doing it that way and it was very awkward. There's a lot of return values that have different meaning depending on what the function ended up doing. I'll make a change in a dedicated commit so you can see for yourself.